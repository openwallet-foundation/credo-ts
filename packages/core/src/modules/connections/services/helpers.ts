import type { Routing } from './ConnectionService'
import type { AgentContext } from '../../../agent'
import type { ResolvedDidCommService } from '../../didcomm'
import type { DidDocument, PeerDidNumAlgo } from '../../dids'
import type { DidDoc, PublicKey } from '../models'

import { Key, KeyType } from '../../../crypto'
import { CredoError } from '../../../error'
import {
  IndyAgentService,
  DidCommV1Service,
  DidDocumentBuilder,
  getEd25519VerificationKey2018,
  DidRepository,
  DidsApi,
  createPeerDidDocumentFromServices,
  DidDocumentRole,
} from '../../dids'
import { didDocumentJsonToNumAlgo1Did } from '../../dids/methods/peer/peerDidNumAlgo1'
import { EmbeddedAuthentication } from '../models'

export function convertToNewDidDocument(didDoc: DidDoc): DidDocument {
  const didDocumentBuilder = new DidDocumentBuilder('')

  const oldIdNewIdMapping: { [key: string]: string } = {}

  didDoc.authentication.forEach((auth) => {
    const { publicKey: pk } = auth

    // did:peer did documents can only use referenced keys.
    if (pk.type === 'Ed25519VerificationKey2018' && pk.value) {
      const ed25519VerificationMethod = convertPublicKeyToVerificationMethod(pk)

      const oldKeyId = normalizeId(pk.id)
      oldIdNewIdMapping[oldKeyId] = ed25519VerificationMethod.id
      didDocumentBuilder.addAuthentication(ed25519VerificationMethod.id)

      // Only the auth is embedded, we also need to add the key to the verificationMethod
      // for referenced authentication this should already be the case
      if (auth instanceof EmbeddedAuthentication) {
        didDocumentBuilder.addVerificationMethod(ed25519VerificationMethod)
      }
    }
  })

  didDoc.publicKey.forEach((pk) => {
    if (pk.type === 'Ed25519VerificationKey2018' && pk.value) {
      const ed25519VerificationMethod = convertPublicKeyToVerificationMethod(pk)

      const oldKeyId = normalizeId(pk.id)
      oldIdNewIdMapping[oldKeyId] = ed25519VerificationMethod.id
      didDocumentBuilder.addVerificationMethod(ed25519VerificationMethod)
    }
  })

  // FIXME: we reverse the didCommServices here, as the previous implementation was wrong
  // and we need to keep the same order to not break the did creation process.
  // When we implement the migration to did:peer:2 and did:peer:3 according to the
  // RFCs we can change it.
  didDoc.didCommServices.reverse().forEach((service) => {
    const serviceId = normalizeId(service.id)

    // For didcommv1, we need to replace the old id with the new ones
    if (service instanceof DidCommV1Service) {
      const recipientKeys = service.recipientKeys.map((keyId) => {
        const oldKeyId = normalizeId(keyId)
        return oldIdNewIdMapping[oldKeyId]
      })

      service = new DidCommV1Service({
        id: serviceId,
        recipientKeys,
        serviceEndpoint: service.serviceEndpoint,
        routingKeys: service.routingKeys,
        accept: service.accept,
        priority: service.priority,
      })
    } else if (service instanceof IndyAgentService) {
      service = new IndyAgentService({
        id: serviceId,
        recipientKeys: service.recipientKeys,
        serviceEndpoint: service.serviceEndpoint,
        routingKeys: service.routingKeys,
        priority: service.priority,
      })
    }

    didDocumentBuilder.addService(service)
  })

  const didDocument = didDocumentBuilder.build()

  const peerDid = didDocumentJsonToNumAlgo1Did(didDocument.toJSON())
  didDocument.id = peerDid

  return didDocument
}

function normalizeId(fullId: string): `#${string}` {
  // Some old dids use `;` as the delimiter for the id. If we can't find a `#`
  // and a `;` exists, we will parse everything after `;` as the id.
  if (!fullId.includes('#') && fullId.includes(';')) {
    const [, ...ids] = fullId.split(';')

    return `#${ids.join(';')}`
  }

  const [, ...ids] = fullId.split('#')
  return `#${ids.length ? ids.join('#') : fullId}`
}

function convertPublicKeyToVerificationMethod(publicKey: PublicKey) {
  if (!publicKey.value) {
    throw new CredoError(`Public key ${publicKey.id} does not have value property`)
  }
  const publicKeyBase58 = publicKey.value
  const ed25519Key = Key.fromPublicKeyBase58(publicKeyBase58, KeyType.Ed25519)
  return getEd25519VerificationKey2018({
    id: `#${publicKeyBase58.slice(0, 8)}`,
    key: ed25519Key,
    controller: '#id',
  })
}

export function routingToServices(routing: Routing): ResolvedDidCommService[] {
  return routing.endpoints.map((endpoint, index) => ({
    id: `#inline-${index}`,
    serviceEndpoint: endpoint,
    recipientKeys: [routing.recipientKey],
    routingKeys: routing.routingKeys,
  }))
}

export async function getDidDocumentForCreatedDid(agentContext: AgentContext, did: string) {
  const didRecord = await agentContext.dependencyManager.resolve(DidRepository).findCreatedDid(agentContext, did)

  if (!didRecord?.didDocument) {
    throw new CredoError(`Could not get DidDocument for created did ${did}`)
  }
  return didRecord.didDocument
}

/**
 * Asserts that the keys we are going to use for creating a did document haven't already been used in another did document
 * Due to how DIDComm v1 works (only reference the key not the did in encrypted message) we can't have multiple dids containing
 * the same key as we won't know which did (and thus which connection) a message is intended for.
 */
export async function assertNoCreatedDidExistsForKeys(agentContext: AgentContext, recipientKeys: Key[]) {
  const didRepository = agentContext.dependencyManager.resolve(DidRepository)
  const recipientKeyFingerprints = recipientKeys.map((key) => key.fingerprint)

  const didsForServices = await didRepository.findByQuery(agentContext, {
    role: DidDocumentRole.Created,

    // We want an $or query so we query for each key individually, not one did document
    // containing exactly the same keys as the did document we are trying to create
    $or: recipientKeyFingerprints.map((fingerprint) => ({
      recipientKeyFingerprints: [fingerprint],
    })),
  })

  if (didsForServices.length > 0) {
    const allDidRecipientKeys = didsForServices.flatMap((did) => did.getTags().recipientKeyFingerprints ?? [])
    const matchingFingerprints = allDidRecipientKeys.filter((f) => recipientKeyFingerprints.includes(f))
    throw new CredoError(
      `A did already exists for some of the keys in the provided services. DIDComm v1 uses key based routing, and therefore it is not allowed to re-use the same key in multiple did documents for DIDComm. If you use the same 'routing' object for multiple invitations, instead provide an 'invitationDid' to the create invitation method. The following fingerprints are already in use: ${matchingFingerprints.join(
        ','
      )}`
    )
  }
}

export async function createPeerDidFromServices(
  agentContext: AgentContext,
  services: ResolvedDidCommService[],
  numAlgo: PeerDidNumAlgo
) {
  const didsApi = agentContext.dependencyManager.resolve(DidsApi)

  // Create did document without the id property
  const didDocument = createPeerDidDocumentFromServices(services)

  // Assert that the keys we are going to use for creating a did document haven't already been used in another did document
  await assertNoCreatedDidExistsForKeys(agentContext, didDocument.recipientKeys)

  // Register did:peer document. This will generate the id property and save it to a did record
  const result = await didsApi.create({
    method: 'peer',
    didDocument,
    options: {
      numAlgo,
    },
  })

  if (result.didState?.state !== 'finished') {
    throw new CredoError(`Did document creation failed: ${JSON.stringify(result.didState)}`)
  }

  return result.didState.didDocument
}
