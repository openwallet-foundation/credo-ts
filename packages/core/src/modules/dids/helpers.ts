import { Key, KeyType } from '../../crypto'
import { isDid } from '../../utils'

import { DidKey } from './methods/key'

export function isDidKey(key: string) {
  return isDid(key, 'key')
}

export function didKeyToVerkey(key: string) {
  if (isDidKey(key)) {
    const publicKeyBase58 = DidKey.fromDid(key).key.publicKeyBase58
    return publicKeyBase58
  }
  return key
}

export function verkeyToDidKey(key: string) {
  if (isDidKey(key)) return key
  const publicKeyBase58 = key
  const ed25519Key = Key.fromPublicKeyBase58(publicKeyBase58, KeyType.Ed25519)
  const didKey = new DidKey(ed25519Key)
  return didKey.did
}

export function didKeyToInstanceOfKey(key: string) {
  const didKey = DidKey.fromDid(key)
  return didKey.key
}

export function verkeyToInstanceOfKey(verkey: string) {
  const ed25519Key = Key.fromPublicKeyBase58(verkey, KeyType.Ed25519)
  return ed25519Key
}
