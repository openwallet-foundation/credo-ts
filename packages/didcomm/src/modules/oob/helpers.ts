import type { OutOfBandDidCommService } from './domain/OutOfBandDidCommService'

import {
  DidCommV1Service,
  DidDocumentBuilder,
  DidKey,
  didDocumentToNumAlgo2Did,
  createPeerDidDocumentFromServices,
  didKeyToInstanceOfKey,
} from '@credo-ts/core'

// This method is kept to support searching for existing connections created by
// credo-ts <= 0.5.1
// TODO: Remove in 0.6.0 (when ConnectionRecord.invitationDid will be migrated)
export function outOfBandServiceToInlineKeysNumAlgo2Did(service: OutOfBandDidCommService) {
  const didDocument = new DidDocumentBuilder('')
    .addService(
      new DidCommV1Service({
        id: service.id,
        serviceEndpoint: service.serviceEndpoint,
        accept: service.accept,
        recipientKeys: service.recipientKeys.map((recipientKey) => {
          const did = DidKey.fromDid(recipientKey)
          return `${did.did}#${did.key.fingerprint}`
        }),
        // Map did:key:xxx to actual did:key:xxx#123
        routingKeys: service.routingKeys?.map((routingKey) => {
          const did = DidKey.fromDid(routingKey)
          return `${did.did}#${did.key.fingerprint}`
        }),
      })
    )
    .build()

  const did = didDocumentToNumAlgo2Did(didDocument)

  return did
}

export function outOfBandServiceToNumAlgo2Did(service: OutOfBandDidCommService) {
  const didDocument = createPeerDidDocumentFromServices([
    {
      id: service.id,
      recipientKeys: service.recipientKeys.map(didKeyToInstanceOfKey),
      serviceEndpoint: service.serviceEndpoint,
      routingKeys: service.routingKeys?.map(didKeyToInstanceOfKey) ?? [],
    },
  ])

  const did = didDocumentToNumAlgo2Did(didDocument)

  return did
}
