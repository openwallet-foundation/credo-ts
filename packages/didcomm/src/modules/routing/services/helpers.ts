import type { AgentContext, DidDocument } from '@credo-ts/core'

import { MediationRecipientService } from './MediationRecipientService'

export async function getMediationRecordForDidDocument(agentContext: AgentContext, didDocument: DidDocument) {
  const [mediatorRecord] = await agentContext.dependencyManager
    .resolve(MediationRecipientService)
    .findAllMediatorsByQuery(agentContext, {
      recipientKeys: didDocument.recipientKeys.map((key) => key.publicKeyBase58),
    })
  return mediatorRecord
}
