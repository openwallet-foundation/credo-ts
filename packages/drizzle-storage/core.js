module.exports = require('./build/core/bundle.js')
