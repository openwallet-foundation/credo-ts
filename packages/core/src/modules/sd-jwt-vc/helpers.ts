import { AgentContext } from '../../agent'
import { getDomainFromUrl } from '../../utils/domain'
import { DidResolverService, DidsApi, getPublicJwkFromVerificationMethod, parseDid } from '../dids'
import { PublicJwk } from '../kms'
import { X509Certificate } from '../x509'
import { SdJwtVcError } from './SdJwtVcError'
import { SdJwtVcHolderBinding, SdJwtVcIssuer } from './SdJwtVcOptions'

async function resolveSigningPublicJwkFromDidUrl(agentContext: AgentContext, didUrl: string) {
  const dids = agentContext.dependencyManager.resolve(DidsApi)

  const { publicJwk } = await dids.resolveVerificationMethodFromCreatedDidRecord(didUrl)
  return publicJwk
}

async function resolveDidUrl(agentContext: AgentContext, didUrl: string) {
  const didResolver = agentContext.dependencyManager.resolve(DidResolverService)
  const didDocument = await didResolver.resolveDidDocument(agentContext, didUrl)

  return {
    verificationMethod: didDocument.dereferenceKey(didUrl, ['assertionMethod']),
    didDocument,
  }
}

function assertValidX5cJwtIssuer(agentContext: AgentContext, iss: string, leafCertificate: X509Certificate) {
  if (!iss.startsWith('https://') && !(iss.startsWith('http://') && agentContext.config.allowInsecureHttpUrls)) {
    throw new SdJwtVcError('The X509 certificate issuer must be a HTTPS URI.')
  }

  if (!leafCertificate.sanUriNames?.includes(iss) && !leafCertificate.sanDnsNames?.includes(getDomainFromUrl(iss))) {
    throw new SdJwtVcError(
      `The 'iss' claim in the payload does not match a 'SAN-URI' name and the domain extracted from the HTTPS URI does not match a 'SAN-DNS' name in the x5c certificate.`
    )
  }
}

export async function extractKeyFromIssuer(agentContext: AgentContext, issuer: SdJwtVcIssuer, forSigning = false) {
  if (issuer.method === 'did') {
    const parsedDid = parseDid(issuer.didUrl)
    if (!parsedDid.fragment) {
      throw new SdJwtVcError(`didUrl '${issuer.didUrl}' does not contain a '#'. Unable to derive key from did document`)
    }

    let publicJwk: PublicJwk
    if (forSigning) {
      publicJwk = await resolveSigningPublicJwkFromDidUrl(agentContext, issuer.didUrl)
    } else {
      const { verificationMethod } = await resolveDidUrl(agentContext, issuer.didUrl)
      publicJwk = getPublicJwkFromVerificationMethod(verificationMethod)
    }

    const supportedSignatureAlgorithms = publicJwk.supportedSignatureAlgorithms
    if (supportedSignatureAlgorithms.length === 0) {
      throw new SdJwtVcError(`No supported JWA signature algorithms found for key ${publicJwk.jwkTypehumanDescription}`)
    }
    const alg = supportedSignatureAlgorithms[0]

    return {
      alg,
      publicJwk,
      iss: parsedDid.did,
      kid: `#${parsedDid.fragment}`,
    }
  }

  // FIXME: probably need to make the input an x509 certificate so we can attach a key id
  if (issuer.method === 'x5c') {
    const leafCertificate = issuer.x5c[0]
    if (!leafCertificate) {
      throw new SdJwtVcError("Empty 'x5c' array provided")
    }

    // TODO: We don't have an x509 certificate record so we expect the key id to already be set
    if (forSigning && !leafCertificate.publicJwk.hasKeyId) {
      throw new SdJwtVcError("Expected leaf certificate in 'x5c' array to have a key id configured.")
    }

    const publicJwk = leafCertificate.publicJwk
    const supportedSignatureAlgorithms = publicJwk.supportedSignatureAlgorithms
    if (supportedSignatureAlgorithms.length === 0) {
      throw new SdJwtVcError(`No supported JWA signature algorithms found for key ${publicJwk.jwkTypehumanDescription}`)
    }
    const alg = supportedSignatureAlgorithms[0]

    assertValidX5cJwtIssuer(agentContext, issuer.issuer, leafCertificate)

    return {
      publicJwk,
      iss: issuer.issuer,
      x5c: issuer.x5c,
      alg,
    }
  }

  throw new SdJwtVcError("Unsupported credential issuer. Only 'did' and 'x5c' is supported at the moment.")
}

export async function extractKeyFromHolderBinding(
  agentContext: AgentContext,
  holder: SdJwtVcHolderBinding,
  forSigning = false
) {
  if (holder.method === 'did') {
    const parsedDid = parseDid(holder.didUrl)
    if (!parsedDid.fragment) {
      throw new SdJwtVcError(`didUrl '${holder.didUrl}' does not contain a '#'. Unable to derive key from did document`)
    }

    let publicJwk: PublicJwk
    if (forSigning) {
      publicJwk = await resolveSigningPublicJwkFromDidUrl(agentContext, holder.didUrl)
    } else {
      const { verificationMethod } = await resolveDidUrl(agentContext, holder.didUrl)
      publicJwk = getPublicJwkFromVerificationMethod(verificationMethod)
    }

    const supportedSignatureAlgorithms = publicJwk.supportedSignatureAlgorithms
    if (supportedSignatureAlgorithms.length === 0) {
      throw new SdJwtVcError(`No supported JWA signature algorithms found for key ${publicJwk.jwkTypehumanDescription}`)
    }
    const alg = supportedSignatureAlgorithms[0]

    return {
      alg,
      publicJwk,
      cnf: {
        // We need to include the whole didUrl here, otherwise the verifier
        // won't know which did it is associated with
        kid: holder.didUrl,
      },
    }
  }
  if (holder.method === 'jwk') {
    const publicJwk = holder.jwk
    const alg = publicJwk.supportedSignatureAlgorithms[0]

    // If there is no key id configured when signing, we assume this credential was issued before we included key ids
    // and the we use the legacy key id.
    if (forSigning && !publicJwk.hasKeyId) {
      publicJwk.keyId = publicJwk.legacyKeyId
    }

    return {
      alg,
      publicJwk,
      cnf: {
        jwk: publicJwk.toJson(),
      },
    }
  }

  throw new SdJwtVcError("Unsupported credential holder binding. Only 'did' and 'jwk' are supported at the moment.")
}
