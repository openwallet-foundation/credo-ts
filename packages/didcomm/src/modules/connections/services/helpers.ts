import { AgentContext, DidDocumentKey, Kms, PeerDidNumAlgo, ResolvedDidCommService } from '@credo-ts/core'
import type { Routing } from '../../../models'
import type { DidDoc, PublicKey } from '../models'

import {
  CredoError,
  DidCommV1Service,
  DidDocumentBuilder,
  DidDocumentRole,
  DidRepository,
  DidsApi,
  IndyAgentService,
  TypedArrayEncoder,
  createPeerDidDocumentFromServices,
  didDocumentJsonToNumAlgo1Did,
  getEd25519VerificationKey2018,
} from '@credo-ts/core'
import { OutOfBandDidCommService } from '../../oob/domain/OutOfBandDidCommService'
import { OutOfBandInlineServiceKey } from '../../oob/repository/OutOfBandRecord'
import { EmbeddedAuthentication } from '../models'

export function convertToNewDidDocument(didDoc: DidDoc, keys?: DidDocumentKey[]) {
  const didDocumentBuilder = new DidDocumentBuilder('')

  const oldIdNewIdMapping: { [key: string]: string } = {}

  for (const auth of didDoc.authentication) {
    const { publicKey: pk } = auth

    // did:peer did documents can only use referenced keys.
    if (pk.type === 'Ed25519VerificationKey2018' && pk.value) {
      const ed25519VerificationMethod = convertPublicKeyToVerificationMethod(pk)

      const oldKeyId = normalizeId(pk.id)
      oldIdNewIdMapping[oldKeyId] = ed25519VerificationMethod.id
      didDocumentBuilder.addAuthentication(ed25519VerificationMethod.id)

      // Only the auth is embedded, we also need to add the key to the verificationMethod
      // for referenced authentication this should already be the case
      if (auth instanceof EmbeddedAuthentication) {
        didDocumentBuilder.addVerificationMethod(ed25519VerificationMethod)
      }
    }
  }

  for (const pk of didDoc.publicKey) {
    if (pk.type === 'Ed25519VerificationKey2018' && pk.value) {
      const ed25519VerificationMethod = convertPublicKeyToVerificationMethod(pk)

      const oldKeyId = normalizeId(pk.id)
      oldIdNewIdMapping[oldKeyId] = ed25519VerificationMethod.id
      didDocumentBuilder.addVerificationMethod(ed25519VerificationMethod)
    }
  }

  // FIXME: we reverse the didCommServices here, as the previous implementation was wrong
  // and we need to keep the same order to not break the did creation process.
  // When we implement the migration to did:peer:2 and did:peer:3 according to the
  // RFCs we can change it.

  for (let service of didDoc.didCommServices.reverse()) {
    const serviceId = normalizeId(service.id)

    // For didcommv1, we need to replace the old id with the new ones
    if (service instanceof DidCommV1Service) {
      const recipientKeys = service.recipientKeys.map((keyId) => {
        const oldKeyId = normalizeId(keyId)
        return oldIdNewIdMapping[oldKeyId]
      })

      service = new DidCommV1Service({
        id: serviceId,
        recipientKeys,
        serviceEndpoint: service.serviceEndpoint,
        routingKeys: service.routingKeys,
        accept: service.accept,
        priority: service.priority,
      })
    } else if (service instanceof IndyAgentService) {
      service = new IndyAgentService({
        id: serviceId,
        recipientKeys: service.recipientKeys,
        serviceEndpoint: service.serviceEndpoint,
        routingKeys: service.routingKeys,
        priority: service.priority,
      })
    }

    didDocumentBuilder.addService(service)
  }

  const didDocument = didDocumentBuilder.build()

  const peerDid = didDocumentJsonToNumAlgo1Did(didDocument.toJSON())
  didDocument.id = peerDid

  return {
    didDocument,
    keys: keys?.map((key) => ({
      ...key,
      didDocumentRelativeKeyId: oldIdNewIdMapping[key.didDocumentRelativeKeyId],
    })),
  }
}

function normalizeId(fullId: string): `#${string}` {
  // Some old dids use `;` as the delimiter for the id. If we can't find a `#`
  // and a `;` exists, we will parse everything after `;` as the id.
  if (!fullId.includes('#') && fullId.includes(';')) {
    const [, ...ids] = fullId.split(';')

    return `#${ids.join(';')}`
  }

  const [, ...ids] = fullId.split('#')
  return `#${ids.length ? ids.join('#') : fullId}`
}

function convertPublicKeyToVerificationMethod(publicKey: PublicKey) {
  if (!publicKey.value) {
    throw new CredoError(`Public key ${publicKey.id} does not have value property`)
  }
  const publicKeyBase58 = publicKey.value
  const ed25519Key = Kms.PublicJwk.fromPublicKey({
    kty: 'OKP',
    crv: 'Ed25519',
    publicKey: TypedArrayEncoder.fromBase58(publicKeyBase58),
  })
  return getEd25519VerificationKey2018({
    id: `#${publicKeyBase58.slice(0, 8)}`,
    publicJwk: ed25519Key,
    controller: '#id',
  })
}

export function routingToServices(routing: Routing): ResolvedDidCommService[] {
  return routing.endpoints.map((endpoint, index) => ({
    id: `#inline-${index}`,
    serviceEndpoint: endpoint,
    recipientKeys: [routing.recipientKey],
    routingKeys: routing.routingKeys,
  }))
}

/**
 * Asserts that the keys we are going to use for creating a did document haven't already been used in another did document
 * Due to how DIDComm v1 works (only reference the key not the did in encrypted message) we can't have multiple dids containing
 * the same key as we won't know which did (and thus which connection) a message is intended for.
 */
export async function assertNoCreatedDidExistsForKeys(agentContext: AgentContext, recipientKeys: Kms.PublicJwk[]) {
  const didRepository = agentContext.dependencyManager.resolve(DidRepository)
  const recipientKeyFingerprints = recipientKeys.map((key) => key.fingerprint)

  const didsForServices = await didRepository.findByQuery(agentContext, {
    role: DidDocumentRole.Created,

    // We want an $or query so we query for each key individually, not one did document
    // containing exactly the same keys as the did document we are trying to create
    $or: recipientKeyFingerprints.map((fingerprint) => ({
      recipientKeyFingerprints: [fingerprint],
    })),
  })

  if (didsForServices.length > 0) {
    const allDidRecipientKeys = didsForServices.flatMap((did) => did.getTags().recipientKeyFingerprints ?? [])
    const matchingFingerprints = allDidRecipientKeys.filter((f) => recipientKeyFingerprints.includes(f))
    throw new CredoError(
      `A did already exists for some of the keys in the provided services. DIDComm v1 uses key based routing, and therefore it is not allowed to re-use the same key in multiple did documents for DIDComm. If you use the same 'routing' object for multiple invitations, instead provide an 'invitationDid' to the create invitation method. The following fingerprints are already in use: ${matchingFingerprints.join(
        ','
      )}`
    )
  }
}

export async function createPeerDidFromServices(
  agentContext: AgentContext,
  services: ResolvedDidCommService[],
  numAlgo: PeerDidNumAlgo
) {
  const didsApi = agentContext.dependencyManager.resolve(DidsApi)

  // Create did document without the id property
  const { didDocument, keys } = createPeerDidDocumentFromServices(services, true)

  // Assert that the keys we are going to use for creating a did document haven't already been used in another did document
  await assertNoCreatedDidExistsForKeys(agentContext, didDocument.recipientKeys)

  // Register did:peer document. This will generate the id property and save it to a did record
  const result = await didsApi.create({
    method: 'peer',
    didDocument,
    options: {
      numAlgo,
      keys,
    },
  })

  if (result.didState?.state !== 'finished') {
    throw new CredoError(`Did document creation failed: ${JSON.stringify(result.didState)}`)
  }

  // FIXME: didApi.create should return the did document
  return didsApi.resolveCreatedDidDocumentWithKeys(result.didState.did)
}

export function getResolvedDidcommServiceWithSigningKeyId(
  outOfBandDidcommService: OutOfBandDidCommService,
  /**
   * Optional keys for the inline services
   */
  inlineServiceKeys?: OutOfBandInlineServiceKey[]
) {
  const resolvedService = outOfBandDidcommService.resolvedDidCommService

  // Make sure the key id is set for service keys
  for (const recipientKey of resolvedService.recipientKeys) {
    const kmsKeyId = inlineServiceKeys?.find(
      ({ recipientKeyFingerprint }) => recipientKeyFingerprint === recipientKey.fingerprint
    )?.kmsKeyId

    recipientKey.keyId = kmsKeyId ?? recipientKey.legacyKeyId
  }

  return resolvedService
}
