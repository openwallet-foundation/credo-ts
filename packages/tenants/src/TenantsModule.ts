import type { Constructor, DependencyManager, EmptyModuleMap, Module, ModulesMap, Update } from '@credo-ts/core'
import type { TenantsModuleConfigOptions } from './TenantsModuleConfig'

import { AgentConfig, InjectionSymbols } from '@credo-ts/core'

import { TenantsApi } from './TenantsApi'
import { TenantsModuleConfig } from './TenantsModuleConfig'
import { TenantAgentContextProvider } from './context/TenantAgentContextProvider'
import { TenantSessionCoordinator } from './context/TenantSessionCoordinator'
import { TenantRepository, TenantRoutingRepository } from './repository'
import { TenantRecordService } from './services'
import { updateTenantsModuleV0_4ToV0_5 } from './updates/0.4-0.5'

export class TenantsModule<AgentModules extends ModulesMap = EmptyModuleMap> implements Module {
  public readonly config: TenantsModuleConfig

  public readonly api: Constructor<TenantsApi<AgentModules>> = TenantsApi

  public constructor(config?: TenantsModuleConfigOptions) {
    this.config = new TenantsModuleConfig(config)
  }

  /**
   * Registers the dependencies of the tenants module on the dependency manager.
   */
  public register(dependencyManager: DependencyManager) {
    // Warn about experimental module
    dependencyManager
      .resolve(AgentConfig)
      .logger.warn(
        "The '@credo-ts/tenants' module is experimental and could have unexpected breaking changes. When using this module, make sure to use strict versions for all @credo-ts packages."
      )

    // Api
    // NOTE: this is a singleton because tenants can't have their own tenants. This makes sure the tenants api is always used in the root agent context.
    dependencyManager.registerSingleton(TenantsApi)

    // Config
    dependencyManager.registerInstance(TenantsModuleConfig, this.config)

    // Services
    dependencyManager.registerSingleton(TenantRecordService)

    // Repositories
    dependencyManager.registerSingleton(TenantRepository)
    dependencyManager.registerSingleton(TenantRoutingRepository)

    dependencyManager.registerSingleton(InjectionSymbols.AgentContextProvider, TenantAgentContextProvider)
    dependencyManager.registerSingleton(TenantSessionCoordinator)
  }

  public updates = [
    {
      fromVersion: '0.4',
      toVersion: '0.5',
      doUpdate: updateTenantsModuleV0_4ToV0_5,
    },
  ] satisfies Update[]
}
