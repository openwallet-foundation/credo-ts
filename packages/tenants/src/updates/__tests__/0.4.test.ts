import { DependencyManager, InjectionSymbols, Agent, UpdateAssistant, utils } from '@credo-ts/core'
import { MediatorRoutingRecord } from '@credo-ts/didcomm'
import { agentDependencies } from '@credo-ts/node'
import { readFileSync } from 'fs'
import path from 'path'

import { InMemoryStorageService } from '../../../../../tests/InMemoryStorageService'
import { RegisteredAskarTestWallet } from '../../../../askar/tests/helpers'
import { TenantsModule } from '../../TenantsModule'

// Backup date / time is the unique identifier for a backup, needs to be unique for every test
const backupDate = new Date('2023-11-23T22:50:20.522Z')
jest.useFakeTimers().setSystemTime(backupDate)

describe('UpdateAssistant | Tenants | v0.4 - v0.5', () => {
  it(`should correctly update the tenant records`, async () => {
    // We need to mock the uuid generation to make sure we generate consistent uuids for the new records created.
    let uuidCounter = 1
    const uuidSpy = jest.spyOn(utils, 'uuid').mockImplementation(() => `${uuidCounter++}-4e4f-41d9-94c4-f49351b811f1`)

    const tenantRecordsString = readFileSync(path.join(__dirname, '__fixtures__/tenants-no-label-tag-0.4.json'), 'utf8')

    const dependencyManager = new DependencyManager()
    const storageService = new InMemoryStorageService()
    dependencyManager.registerInstance(InjectionSymbols.StorageService, storageService)
    // If we register the AskarModule it will register the storage service, but we use in memory storage here
    dependencyManager.registerContextScoped(InjectionSymbols.Wallet, RegisteredAskarTestWallet)

    const agent = new Agent(
      {
        config: {
          label: 'Test Agent',
          walletConfig: {
            id: `Wallet: 0.5 Update Tenants`,
            key: `Key: 0.5 Update Tenants`,
          },
        },
        dependencies: agentDependencies,
        modules: {
          // We need to include the TenantsModule to run the updates
          tenants: new TenantsModule(),
        },
      },
      dependencyManager
    )

    const updateAssistant = new UpdateAssistant(agent, {
      v0_1ToV0_2: {
        mediationRoleUpdateStrategy: 'doNotChange',
      },
    })

    await updateAssistant.initialize()

    // Set storage after initialization. This mimics as if this wallet
    // is opened as an existing wallet instead of a new wallet
    storageService.contextCorrelationIdToRecords = {
      default: {
        records: JSON.parse(tenantRecordsString),
        creationDate: new Date(),
      },
    }

    expect(await updateAssistant.isUpToDate('0.5')).toBe(false)
    expect(await updateAssistant.getNeededUpdates('0.5')).toEqual([
      {
        fromVersion: '0.4',
        toVersion: '0.5',
        doUpdate: expect.any(Function),
      },
    ])

    await updateAssistant.update({ updateToVersion: '0.5' })

    expect(await updateAssistant.isUpToDate('0.5')).toBe(true)
    expect(await updateAssistant.getNeededUpdates('0.5')).toEqual([])

    await storageService.deleteById(agent.context, MediatorRoutingRecord, 'MEDIATOR_ROUTING_RECORD')
    expect(storageService.contextCorrelationIdToRecords[agent.context.contextCorrelationId].records).toMatchSnapshot()

    await agent.shutdown()
    await agent.wallet.delete()

    uuidSpy.mockReset()
  })
})
