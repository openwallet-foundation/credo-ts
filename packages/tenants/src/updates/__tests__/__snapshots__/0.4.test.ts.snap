// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateAssistant | Tenants | v0.4 - v0.5 should correctly update the tenant records 1`] = `
{
  "1-4e4f-41d9-94c4-f49351b811f1": {
    "id": "1-4e4f-41d9-94c4-f49351b811f1",
    "tags": {
      "label": "Tenant 1",
      "storageVersion": "0.4",
    },
    "type": "TenantRecord",
    "value": {
      "config": {
        "label": "Tenant 1",
        "walletConfig": {
          "id": "tenant-1-4e4f-41d9-94c4-f49351b811f1",
          "key": "Bs2YXMZ4mRRYSEzHPGVMuGfAD3qt9xwjgbMFNW4wuVKr",
          "keyDerivationMethod": "RAW",
        },
      },
      "createdAt": "2023-11-23T22:50:20.522Z",
      "id": "1-4e4f-41d9-94c4-f49351b811f1",
      "metadata": {},
      "storageVersion": "0.4",
      "updatedAt": "2023-11-23T22:50:20.522Z",
    },
  },
  "2-4e4f-41d9-94c4-f49351b811f1": {
    "id": "2-4e4f-41d9-94c4-f49351b811f1",
    "tags": {
      "label": "Tenant 2",
      "storageVersion": "0.4",
    },
    "type": "TenantRecord",
    "value": {
      "config": {
        "label": "Tenant 2",
        "walletConfig": {
          "id": "tenant-2-4e4f-41d9-94c4-f49351b811f1",
          "key": "5eB1uJwE7c9hkyJ8cba5ziPCZ5GkXNjnhVKh95jBMfou",
          "keyDerivationMethod": "RAW",
        },
      },
      "createdAt": "2023-11-23T22:50:20.522Z",
      "id": "2-4e4f-41d9-94c4-f49351b811f1",
      "metadata": {},
      "storageVersion": "0.4",
      "updatedAt": "2023-11-23T22:50:20.522Z",
    },
  },
  "STORAGE_VERSION_RECORD_ID": {
    "id": "STORAGE_VERSION_RECORD_ID",
    "tags": {},
    "type": "StorageVersionRecord",
    "value": {
      "createdAt": "2023-11-23T22:50:20.522Z",
      "id": "STORAGE_VERSION_RECORD_ID",
      "metadata": {},
      "storageVersion": "0.5",
      "updatedAt": "2023-11-23T22:50:20.522Z",
    },
  },
}
`;
